pkg/operator/nfs/controller_test.go

```diff
@@ -13,170 +13,300 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-const appName = "nfs-server-X"
-
-func TestValidateNFSServerSpec(t *testing.T) {
-
-	// first, test that a good NFSServerSpec is good
-	spec := nfsv1alpha1.NFSServerSpec{
-		Replicas: 1,
-		Exports: []nfsv1alpha1.ExportsSpec{
+func TestNFSServerReconciler_Reconcile(t *testing.T) {
+	expectedStatefulSet := func(scheme *runtime.Scheme, cr *nfsv1alpha1.NFSServer) *appsv1.StatefulSet {
+		sts := newStatefulSetForNFSServer(cr)
+		sts.Spec.Selector = &metav1.LabelSelector{
+			MatchLabels: newLabels(cr),
+		}
+		_ = controllerutil.SetControllerReference(cr, sts, scheme)
+		volumes := []corev1.Volume{
 			{
-				Name: "test",
-				Server: nfsv1alpha1.ServerSpec{
-					AccessMode: "readwrite",
-					Squash:     "none",
+				Name: cr.Name,
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: cr.Name,
+						},
+						Items: []corev1.KeyToPath{
+							{
+								Key:  cr.Name,
+								Path: cr.Name,
+							},
+						},
+						DefaultMode: pointer.Int32Ptr(corev1.ConfigMapVolumeSourceDefaultMode),
+					},
 				},
 			},
-		},
+		}
+		volumeMounts := []corev1.VolumeMount{
+			{
+				Name:      cr.Name,
+				MountPath: nfsConfigMapPath,
+			},
+		}
+		for _, export := range cr.Spec.Exports {
+			shareName := export.Name
+			claimName := export.PersistentVolumeClaim.ClaimName
+			volumes = append(volumes, corev1.Volume{
+				Name: shareName,
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: claimName,
+					},
+				},
+			})
+
+			volumeMounts = append(volumeMounts, corev1.VolumeMount{
+				Name:      shareName,
+				MountPath: "/" + claimName,
+			})
+		}
+		sts.Status.ReadyReplicas = int32(cr.Spec.Replicas)
+		sts.Spec.Template.Spec.Volumes = volumes
+		sts.Spec.Template.Spec.Containers[0].VolumeMounts = volumeMounts
+
+		return sts
 	}
 
-	err := validateNFSServerSpec(spec)
-	assert.Nil(t, err)
+	rr := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      "nfs-server",
+			Namespace: "nfs-server",
+		},
+	}
 
-	// test that AccessMode is invalid
-	spec = nfsv1alpha1.NFSServerSpec{
-		Replicas: 1,
-		Exports: []nfsv1alpha1.ExportsSpec{
-			{
-				Name: "test",
-				Server: nfsv1alpha1.ServerSpec{
-					AccessMode: "badValue",
-					Squash:     "none",
+	type args struct {
+		req ctrl.Request
+	}
+	tests := []struct {
+		name    string
+		args    args
+		cr      *nfsv1alpha1.NFSServer
+		want    ctrl.Result
+		wantErr bool
+	}{
+		{
+			name: "Reconcile NFS Server Should Set Initializing State when State is Empty",
+			args: args{
+				req: rr,
+			},
+			cr: &nfsv1alpha1.NFSServer{
```

Comment:
we can have a helper function that can generate `nfsv1alpha1.NFSServer{}` and update the Exports.

Reply:
addressed, thanks @rohan47